_fcmaescpp/gcldeoptimizer.cpp

```diff
@@ -145,208 +149,219 @@ class GclDeOptimizer {
+			double stopfitness_, double F0_, double CR0_) {
+		// runid used to identify a specific run
+		runid = runid_;
+		// fitness function to minimize
+		fitfun = fitfun_;
+		// Number of objective variables/problem dimension
+		dim = dim_;
+		// Population size
+		if (popsize_ > 0)
+			popsize = popsize_;
+		else
+			popsize = 15 * dim;
+		// termination criteria
+		// maximal number of evaluations allowed.
+		maxEvaluations = maxEvaluations_;
+		// use low value 0 < pbest <= 1 to narrow search.
+		pbest = pbest_;
+		// Limit for fitness value.
+		stopfitness = stopfitness_;
+		F0 = F0_;
+		CR0 = CR0_;
+		// Number of iterations already performed.
+		iterations = 0;
+		bestValue = DBL_MAX;
+		// stop criteria
+		stop = 0;
+		//std::random_device rd;
+		rs = new pcg64(seed_);
+		init();
+	}
+
+	~GclDeOptimizer() {
+		delete rs;
+	}
+
+	double rnd01() {
+		return distr_01(*rs);
+	}
+
+	int rndInt(int max) {
+		return (int) (max * distr_01(*rs));
+	}
+
+	int findNear(vec x, vector<InfoStore> storage) {
+		int index = 0;
+		double minDist = DBL_MAX;
+		for (int j = 0; j < storage.size(); j++) {
+			double distance = (x - storage[j].x).squaredNorm();
+			if (distance < minDist) {
+				index = j;
+				minDist = distance;
+			}
+		}
+		return index;
+	}
+
+	void doOptimize() {
+		int Gr = 2;
+		int gen_stuck = 0;
+		vector<vec> sp;
+		vector<InfoStore> storage;
+
+		int maxIter = maxEvaluations / popsize + 1;
+		double previous_best = DBL_MAX;
+		double CR = CR0;
+		double F = F0;
+
+		// -------------------- Generation Loop --------------------------------
+
+		for (iterations = 1; fitfun->getEvaluations() < maxEvaluations;
+				iterations++) {
+			mat nextX = mat(popX);
+			vec nextY = vec(popY);
+			vec popCR = vec(popsize);
+			vec popF = vec(popsize);
+			vec xb = popX.col(0);
+			for (int p = 0; p < popsize; p++) {
+				vec xi = popX.col(p);
+				int r1, r2, r3;
+				do {
+					r1 = rndInt(popsize);
+				} while (r1 == p);
+				do {
+					r2 = rndInt(int(popsize * pbest));
+				} while (r2 == p || r2 == r1);
+				do {
+					r3 = rndInt(popsize + sp.size());
+				} while (r3 == p || r3 == r2 || r3 == r1);
+				int jr = rndInt(dim);
+				//Produce the CR and F
+				float mu = 1
+						- sqrt(float(iterations / maxIter))
+								* exp(float(-gen_stuck / iterations));
+				if (storage.empty() || iterations < Gr) {
+					CR = normreal(rs, 0.95, 0.01);
+					F = normreal(rs, mu, 1);
+					if (F < 0 || F > 1)
+						F = rnd01();
+				} else {
+					int i = findNear(popX.col(p), storage);
+					CR = storage[i].CR;
+					F = storage[i].F;
+				}
+				popCR[p] = CR;
+				popF[p] = F;
+				vec ui = vec(xi);
+				for (int j = 0; j < dim; j++) {
+					if (j == jr || rnd01() < CR) {
+						if (r3 < popsize)
+							ui[j] = popX(j, r1)
+									+ F * (popX(j, r2) - popX(j, r3));
+						else
+							ui[j] = popX(j, r1)
+									+ F * (popX(j, r2) - sp[r3 - popsize][j]);
+						if (ui[j] > 1 || ui[j] < 0)
+							ui[j] = rnd01();
+					}

```

Comment:
please check, there should be an "else". Something has been missed.

Reply:
In your code it is 

				if (rndreal(0,1) < CR || j == j_rnd)
				{
					if (r3 < pop_size +1)
					sde_child_pop[i].xreal[j] = sde_parent_pop[r1].xreal[j] +
						F*(sde_parent_pop[r2].xreal[j]-sde_parent_pop[r3].xreal[j]);
					else
						sde_child_pop[i].xreal[j] = sde_parent_pop[r1].xreal[j] +
						F*(sde_parent_pop[r2].xreal[j]-sp[r3-pop_size].xreal[j]);
					if (sde_child_pop[i].xreal[j] < lowerBound[j] ||
						sde_child_pop[i].xreal[j] > upperBound[j])
					{
						sde_child_pop[i].xreal[j] =
							rndreal(lowerBound[j], upperBound[j]);
					}
				}
				else
				{
					sde_child_pop[i].xreal[j] = sde_parent_pop[i].xreal[j];
				}

There is no else to the boundary check. 

In your else you do
sde_child_pop[i].xreal[j] = sde_parent_pop[i].xreal[j];

which copies the value from the parent. 

This is missing in my code since

mat nextX = mat(popX);

already copied the whole population to the child. So the old variable settings are already in place, I only have to do something if they change.